osgp/protocol-adapter-oslp/oslp/src/main/java/org/opensmartgridplatform/oslp/OslpDecoder.java

```diff
@@ -42,70 +46,90 @@ public OslpDecoder(final String signature, final String provider) {
+        if (this.state().compareTo(DecodingState.DEVICE_ID) <= 0) {
+            LOGGER.debug("Decoding device id.");
+            this.decodeDeviceId(in);
             this.checkpoint(DecodingState.LENGTH_INDICATOR);
         }
 
-        if (state.compareTo(DecodingState.LENGTH_INDICATOR) <= 0) {
-            this.decodeLengthIndicator(buffer);
+        if (this.state().compareTo(DecodingState.LENGTH_INDICATOR) <= 0) {
+            LOGGER.debug("Decoding length indicator.");
+            this.decodeLengthIndicator(in);
             this.checkpoint(DecodingState.PAYLOAD_MESSAGE);
         }
 
-        if (state.compareTo(DecodingState.PAYLOAD_MESSAGE) <= 0) {
-            this.builder.withPayloadMessage(Oslp.Message.parseFrom(buffer.readBytes(this.length).array()));
+        if (this.state().compareTo(DecodingState.PAYLOAD_MESSAGE) <= 0) {
+            LOGGER.debug("Decoding payload.");
+            this.decodePayload(in);
             this.checkpoint(DecodingState.SECURITY_KEY);
             try {
-                return this.builder.withSignature(this.signature).withProvider(this.provider).build();
+                final OslpEnvelope msg = this.builder.withSignature(this.signature).withProvider(this.provider).build();
+                out.add(msg);
             } finally {
                 this.reset();
             }
         } else {
             // Should not get here
-            throw new UnknownOslpDecodingStateException(state.name());
+            throw new UnknownOslpDecodingStateException(this.state().name());
         }
     }
 
-    private void decodeSecurityKey(final ChannelBuffer buffer) {
-        this.builder.withSecurityKey(buffer.readBytes(OslpEnvelope.SECURITY_KEY_LENGTH).toByteBuffer().array());
+    private void decodeSecurityKey(final ByteBuf buffer) {
+        final byte[] bytes = ByteBufUtil.getBytes(buffer, buffer.readerIndex(), OslpEnvelope.SECURITY_KEY_LENGTH);
+        this.builder.withSecurityKey(bytes);
+        buffer.readerIndex(buffer.readerIndex() + OslpEnvelope.SECURITY_KEY_LENGTH);
     }
 
-    private void decodeSequenceNumber(final ChannelBuffer buffer) {
-        this.builder.withSequenceNumber(buffer.readBytes(OslpEnvelope.SEQUENCE_NUMBER_LENGTH).toByteBuffer().array());
-
+    private void decodeSequenceNumber(final ByteBuf buffer) {
+        final byte[] bytes = ByteBufUtil.getBytes(buffer, buffer.readerIndex(), OslpEnvelope.SEQUENCE_NUMBER_LENGTH);
+        this.builder.withSequenceNumber(bytes);
+        buffer.readerIndex(buffer.readerIndex() + OslpEnvelope.SEQUENCE_NUMBER_LENGTH);
     }
 
-    private void decodeDeviceId(final ChannelBuffer buffer) {
-        this.builder.withDeviceId(buffer.readBytes(OslpEnvelope.DEVICE_ID_LENGTH + OslpEnvelope.MANUFACTURER_ID_LENGTH)
-                .toByteBuffer().array());
+    private void decodeDeviceId(final ByteBuf buffer) {
+        final int length = OslpEnvelope.DEVICE_ID_LENGTH + OslpEnvelope.MANUFACTURER_ID_LENGTH;
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Rename "length" which hides the field declared at line 38. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AHiddenFieldCheck)